### PR TITLE
providers/aws: support slow draining

### PR DESCRIPTION
**Pull request by pcarrier**
Useful when doing blue/green deploys.

### CODE REVIEW

File: builtin/providers/aws/resource_aws_autoscaling_group.go [hunk 1]
```diff
@@ -518,20 +537,34 @@ func resourceAwsAutoscalingGroupDrain(d *schema.ResourceData, meta interface{})
...
+		}
+		delayT := time.Duration(delay.(int)) * time.Second
+		t := time.Now()
+		count := *g.DesiredCapacity
+		for count > 0 {
+			count--
+			if err := resourceAwsAutoscalingScaleDownToCount(d, count, conn); err != nil {
+				return err
+			}
+			now := time.Now()
```
Comment (glasser) [inline builtin/providers/aws/resource_aws_autoscaling_group.go]: Do we really want to sleep even after the final "scale to 0"?  I think maybe there should be a `if count == 0 { break }` here

Comment (pcarrier) [inline builtin/providers/aws/resource_aws_autoscaling_group.go]: Good catch. Addressed.

**Comment by glasser** [comment]
Is there anything we can do to help this get merged?

**Comment by jen20** [comment]
Hi @pcarrier! Sorry this has taken so long to get around to. Could you describe a bit more about when this is useful, and whether or not there is a good way to test it?

**Comment by glasser** [comment]
We (I work with Pierre) use it when replacing an ASG with another one which
is a proxy tier which services long term user connections. Doing this lets
us break and transfer connections (leading to reconnects) from one proxy at
a time rather than killing the whole asg at once at AWS autoscaling's pace.

It's imaginable that ASG has a built in way to control down-scaling rate
but we don't know about it.

**Comment by iiro** [comment]
@glasser are you doing the blue/green using this method: https://aws.amazon.com/blogs/aws/attach-and-detach-elastic-load-balancers-from-auto-scaling-groups/ ? We're using this... With connection draining- option enabled in the ELB. Not completely sure are we completely protected for connection breaking when detaching the old ASG from ELB...

**Comment by glasser** [comment]
@iiro Roughly, but as our users tend to have long running connections (websockets) a longer period between forced disconnects works better.

**Comment by glasser** [comment]
@jen20 Oops, I forgot to answer the question about testing. Setting this option has no effect on the final state after applying an update, just on how long it takes to get there (and on what intermediate states an observer looking directly at the ASG parameters would see while the update executes). I'm not sure if there's an easy way to observe this in the standard acceptance testing framework, which as far as I know mostly gives you a way to look at state between tests? One could add a test that uses the parameter and at least shows that it properly scales you down eventually, though. What do you think?

**Comment by glasser** [comment]
Please don't merge this at this time. There's a logic bug in calculating the delete delay that causes every other delay to be short. Moreover, we're reworking this to actually wait for the ASG to scale down rather than calculate the delay from the beginning of the last operation.

**Comment by glasser** [comment]
I've updated this to wait for each machine to actually be gone before moving on. (And fixed the logic bug referenced above.) It's now worth merging.

**Comment by shaharmor** [comment]
Any update on this?

**Comment by glasser** [comment]
We have been using it successfully in production since March, though there were some serious issues with it that we fixed in September. We have been using it successfully again since then. However, we have no control over the merge process.

(It also would be nice if the same option could be applied when updating an ASG to make it smaller and not just when deleting it, but I think it's a worthwhile change without that.)

**Comment by shaharmor** [comment]
@glasser is the code in this PR after the September fixes?
Also, how does this work in combination with Application Load Balancer & Target Groups?

In Target Groups you can configure "Connection Draining"  that a terminated (Or is it only for DeRegistered) instances should wait in "draining" mode before terminated.
How does this handle these situations?

**Comment by glasser** [comment]
Yes, this PR was updated in September.

We don't use ALB or target groups.

**Comment by glasser** [comment]
Updated and fixed merge conflicts.

We have been using this in production for over a year. It's super useful! I'd love it if others could use it!

**Comment by apparentlymart** [comment]
Hello @pcarrier, and thanks for working on this!

As part of the the Terraform 0.10 release earlier this year, all of the Terraform providers were moved to their own repositories in [the terraform-providers GitHub organization](https://github.com/terraform-providers), and removed from the Terraform Core repository.

Unfortunately due to the fact that new issues and pull requests are being opened constantly, it was not possible for the various provider maintainers to merge all outstanding pull requests before this split, and there is no automatic way to migrate a pull request to a new repository.

As a result, this pull request can sadly no longer be applied as-is, and so I'm going to close it.

If you or someone else has the time and motivation to apply same changes to [the `aws` provider repository](https://github.com/terraform-providers/terraform-provider-aws) and open a new PR there, the maintainers of that provider should be able to review and merge it.

Thanks again for working on this, and sorry it was not able to be merged before the provider repository changes.

**Pull request closed without merging**